lib/logthrsource/logthrsourcedrv.c

```diff
@@ -0,0 +1,324 @@
+static LogPipe *
+log_threaded_source_worker_logpipe(LogThreadedSourceWorker *self)
+{
+  return &self->super.super;
+}
+
+static void
+log_threaded_source_worker_set_options(LogThreadedSourceWorker *self, LogThreadedSourceDriver *control,
+                                       LogThreadedSourceWorkerOptions *options,
+                                       const gchar *stats_id, const gchar *stats_instance)
+{
+  /* TODO: support position tracking */
+  log_source_set_options(&self->super, &options->super, stats_id, stats_instance, TRUE, FALSE,
+                         control->super.super.super.expr_node);
+
+  self->control = control;
+}
+
+void
+log_threaded_source_worker_options_defaults(LogThreadedSourceWorkerOptions *options)
+{
+  log_source_options_defaults(&options->super);
+}
+
+void
+log_threaded_source_worker_options_init(LogThreadedSourceWorkerOptions *options, GlobalConfig *cfg,
+                                        const gchar *group_name)
+{
+  log_source_options_init(&options->super, cfg, group_name);
+}
+
+void
+log_threaded_source_worker_options_destroy(LogThreadedSourceWorkerOptions *options)
+{
+  log_source_options_destroy(&options->super);
+}
+
+/* The wakeup lock must be held before calling this function. */
+static void
+log_threaded_source_suspend(LogThreadedSourceDriver *self)
+{
+  LogThreadedSourceWorker *worker = self->worker;
+
+  worker->wakeup_cond.awoken = FALSE;
+  while (!worker->wakeup_cond.awoken)
+    g_cond_wait(worker->wakeup_cond.cond, worker->wakeup_cond.lock);
+}
+
+static void
+log_threaded_source_wakeup(LogThreadedSourceDriver *self)
+{
+  LogThreadedSourceWorker *worker = self->worker;
+
+  g_mutex_lock(worker->wakeup_cond.lock);
+  worker->wakeup_cond.awoken = TRUE;
+  g_cond_signal(worker->wakeup_cond.cond);
+  g_mutex_unlock(worker->wakeup_cond.lock);
+}
+
+static void
+log_threaded_source_worker_run(LogThreadedSourceWorker *self)
+{
+  msg_debug("Worker thread started", evt_tag_str("driver", self->control->super.super.id));
+
+  self->run(self->control);
+
+  msg_debug("Worker thread finished", evt_tag_str("driver", self->control->super.super.id));
+}
+
+static void
+log_threaded_source_worker_request_exit(LogThreadedSourceWorker *self)
+{
+  /* TODO force exit after timeout */
+  msg_debug("Requesting worker thread exit", evt_tag_str("driver", self->control->super.super.id));
+  self->request_exit(self->control);
+  log_threaded_source_wakeup(self->control);
+}
+
+static void
+_worker_wakeup(LogSource *s)
+{
+  LogThreadedSourceWorker *self = (LogThreadedSourceWorker *) s;
+
+  self->wakeup(self->control);
+}
+
+static void
+_start_worker_thread(gint type, gpointer data)
+{
+  LogThreadedSourceWorker *self =  (LogThreadedSourceWorker *) data;
+
+  main_loop_create_worker_thread((WorkerThreadFunc) log_threaded_source_worker_run,
+                                 (WorkerExitNotificationFunc) log_threaded_source_worker_request_exit,
+                                 self, &self->options);
+}
```

Comment:
Yeah, but I think this is not required on the source side, because the source thread itself generates new messages after it is started.

I tried to find a similar race that you had on the destination side, but I could not.